src/scripts/generate_binary_ops.py

```diff
@@ -0,0 +1,305 @@
+        op2 = random.randint(_constants.uint32min, _constants.uint32max)
+        rnum = unsigned32(op2)
+        shiftcount = rnum & 0x1f
+        sum = op1 << shiftcount
+        if sum > _constants.bitmax or sum < _constants.bitmin: # Operand 2 must depend on the length of operand 1
+            if op1 != 0:
+                op2 = (_constants.bitmax_exposant - 1) - int.bit_length(op1)
+                op2 = max(op2, 0) # Operand 2 also cant go below zero
+                rnum = unsigned32(op2)
+                shiftcount = rnum & 0x1f
+                sum = op1 << shiftcount
+        return {'oper': '<<', 'op2': op2, 'sum': sum}
+
+    # Bitshifting to right
+    # Python only can handle shiftcount between 0-31, but smaller and bigger numbers
+    # are also can be simulated with this method
+    def bitrightshift(self, op1, op2):
+        op2 = random.randint(_constants.bitmin, _constants.bitmax)
+        sum = int(hex((op1 >> ((op2 % (_constants.bitmax_exposant + 1)) & 0x1f))), 16)
+        return {'oper': '>>', 'op2': op2, 'sum': sum}
+
+    # Unsigned right bitshifting
+    def bitunsignedrightshift(self, op1, op2):
+        op2 = random.randint(_constants.uint32min, _constants.uint32max)
+        lnum = unsigned32(op1)
+        rnum = unsigned32(op2)
+        shiftcount = rnum & 0x1f
+        sum = ((lnum >> shiftcount) % _constants.uint32max)
+        return {'oper': '>>>', 'op2': op2, 'sum': sum}
+
+    # Bitwise and
+    def bitand(self, op1, op2):
+        sum = op1 & op2
+        return {'oper': '&', 'op2': op2, 'sum': sum}
+
+    # Bitwise or
+    def bitor(self, op1, op2):
+        sum = op1 | op2
+        return {'oper': '|', 'op2': op2, 'sum': sum}
+
+    # Bitwise xor (exclusive or)
+    def bitxor(self, op1, op2):
+        sum = op2 ^ op1
+        return {'oper': '^', 'op2': op2, 'sum': sum}
+
+    # Setting maximum and minimum depending on operation type
+    # 1: arithmetic, 2: bitwise (excluding unsigned right shift) 3: unsigned right shift
+    def set_min_max(self, calc):
+        if calc == self._ARITH:
+            MINNUM = _constants.min
+            MAXNUM = _constants.max
+        elif calc == self._BIT:
+            MINNUM = _constants.bitmin
+            MAXNUM = _constants.bitmax
+        else:
+            MINNUM = _constants.uint32min
+            MAXNUM = _constants.uint32max
+        return {'MIN' : MINNUM, 'MAX': MAXNUM}
+
+    # Converting expression, expected value, false value in a function call to string format
+    def string_creator(self, MINNUM, MAXNUM):
+        last = self.operands[0]
+        expression = ""
+        expression += ("%s%s" % ("(" * (self.num_operands - 1), self.TEST_VALUE_VAR_NAME.format(NUMBER = 1)))
+        for i in range(self.options.operand_count - 1):
+            result = self.operators_list[self.operators[i]](last, self.operands[i + 1])
+            expression += (" %s %s)" % (result["oper"], self.TEST_VALUE_VAR_NAME.format(NUMBER = (i + 2))))
```

Comment:
The additional whitespace character is needed before the first %s?

Reply:
It is needed, because this way the expression looks like
```
(((((((((t1 | t2) >> t3) | t4) & t5) >> t6) | t7) & t8) | t9) << t10)
```
Without that there wouldn't be spaces before the operators.